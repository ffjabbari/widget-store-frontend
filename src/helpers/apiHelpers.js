/** API Helper functions. */

// local imports
import { asyncAPIRequest } from '../hooks/apiHook';

/**
 * updateUserProfile()
 *
 * Make patch request to update a user profile. Used by CustomerAddressForm
 * to allow updating customer address of signed in users.
 *
 * @param {object} values - form values
 * @param {object} user - user object
 * @param {token} token - user token
 * @param {function} setResponseError - state setter for error display
 * @param {any} defaultAddress - Falsey or obj with address components
 * @param {function} setFormDisabled - state setter to hide form after update
 *
 * returns: true if an error happened else undefined
 */

export const updateUserProfile = async (
  values,
  user,
  token,
  setResponseError,
  defaultAddress,
  setFormDisabled
) => {
  const userUpdateData = { ...values };
  // remove email to dis-allow email updating from here
  delete userUpdateData['email'];
  delete userUpdateData['user_id'];

  // remomve key/value pairs having a value of empty string from userUpdateData
  Object.entries(userUpdateData).forEach(([key, value]) => {
    if (value === '') delete userUpdateData[key];
  });

  try {
    // make patch request
    const resp = await asyncAPIRequest(
      `/users/${user.username}`,
      'patch',
      userUpdateData,
      { _token: token }
    );
    if (resp.error) {
      setResponseError(resp.error.response.data.message);
      return true;
    } else {
      defaultAddress.current = values;
      setFormDisabled(true);
    }
  } catch (error) {
    setResponseError(error.message);
    return true;
  }
};
