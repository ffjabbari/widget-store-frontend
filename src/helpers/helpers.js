import { Redirect } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
