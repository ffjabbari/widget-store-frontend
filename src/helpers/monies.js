import Decimal from 'decimal.js';

const SF_SALES_TAX = new Decimal('0.085');

// calculate discounted price
export function calculateDiscountPrice(product) {
  if (product.discount === '0.00') return product.price;
  const discount = new Decimal(product.discount),
    price = new Decimal(product.price);
  return price.minus(price.times(discount)).toFixed(2);
}

/** Calculate discount as a percent. */
export function calculateDiscountPercent(product) {
  const discount = new Decimal(product.discount);
  return discount.times(100).toFixed(0);
}

/** Return discounted price and discount as a percent */
export function getDiscountData(product) {
  return {
    discountPrice: calculateDiscountPrice(product),
    percent: calculateDiscountPercent(product),
  };
}

/** Calculate tax if customer is in California */
export function calculateTax(subtotal, state) {
  if (!/^ca(lifornia)?$/i.test(state)) return '0.00';
  subtotal = new Decimal(subtotal);
  return subtotal.times(SF_SALES_TAX).toFixed(2);
}

/** Calculate tax plus subtotal plus shipping cost if provided */
export function calculateTotal(subtotal, state, shippingCost) {
  const tax = calculateTax(subtotal, state);
  subtotal = new Decimal(subtotal);
  shippingCost = shippingCost || 0;
  return subtotal.plus(tax).plus(shippingCost).toFixed(2);
}
