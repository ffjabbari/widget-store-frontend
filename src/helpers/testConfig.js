/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'A good remote.',
    discount: '0.15',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    rating: null,
    quantity: 1,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: '',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
    set_default: false,
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        date_added: '2021-01-24T08:00:00.000Z',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 2,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };
};
