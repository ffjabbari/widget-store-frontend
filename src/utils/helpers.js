import { Redirect, Link } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * getPathRoot()
 *
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

/**
 * <FramerRedirect to="" />
 *
 * Component to force react-router-dom redirect when framer-motion
 * prevents a redirect from executing.
 *
 * If regular <Redirect /> fails in this environment use this redirect instead.
 *
 * @param {any} props
 *
 * returns: motion.div wrapped redirect component
 */
export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);
export const FramerLink = props => (
  <motion.div exit="undefined">
    <Link {...props} />
  </motion.div>
);

/**
 * animateVariant()
 *
 * Animation function to turn on an animation state then turn it back
 * off after a delay. The state will be set "active" then back to "default".
 *
 * If inital state provided by setterFunct is false this function will skip
 * setting the animation state "active" instead set it to "default".
 *
 * If a callback is provided it will be called with any arguments
 * after setting state back to "default".
 *
 *
 * @param {function} setterFunct - useState setter funct to set animation state
 * @param {integer} delay - delay of setterFunct being called
 * @param {funct} callback - function to call upon completion
 * @param {array} args - array of arguments for callback function
 */

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  // after delay set back to "default"
  return setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
