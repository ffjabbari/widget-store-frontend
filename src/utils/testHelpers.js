import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { testStore } from '../redux/store/reduxStore';

export const renderWithStore = component =>
  render(
    <Provider store={testStore}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
