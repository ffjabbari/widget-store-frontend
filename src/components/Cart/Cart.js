import { useState } from 'react';
import { useSelector } from 'react-redux';
import { useHistory, useLocation } from 'react-router-dom';
import styled from 'styled-components';
import { motion } from 'framer-motion';

// local imports
import CartHeader from './Cart_components/CartHeader';
import CartFooter from './Cart_components/CartFooter';
import CartList from '../common/CartList/CartList';
import { getPathRoot } from '../../helpers/helpers';
import './Cart.css';

// background to fade page behind open cart
const ScreenBackground = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right 0;
  background: black;
  z-index: 99;
`;

const SideCart = styled.div`
  position: fixed;
  top: 0px;
  bottom: 0;
  width: 100vw;
  background: var(--light);
  overflow-y: auto;
  z-index: 100;
  border-right: 1px solid var(--secondary);
  border-left: 1px solid var(--secondary);

  @media screen and (min-width: 992px) {
    width: 540px;
  }
`;

function Cart() {
  const history = useHistory();
  const location = useLocation();

  const numCartItems = useSelector(state => state.cart.numCartItems);

  // SideCart animation state
  const [cartAnimationClass, setcartAnimationClass] = useState('animActive');

  const handleClose = () => {
    setcartAnimationClass('animReverse');
    setTimeout(() => history.push(`${getPathRoot(location.pathname)}`), 400);
  };

  // checkout logic function
  const goToCheckout = () => {
    if (numCartItems === 0) return;
    // close side-cart and navigate to checkout
    setcartAnimationClass('animReverse');
    setTimeout(() => history.push('/checkout'), 500);
  };

  return (
    <>
      <ScreenBackground
        onClick={handleClose}
        className={`ScreenBackground ${cartAnimationClass}`}
      />

      <SideCart className={`SideCart ${cartAnimationClass}`}>
        <CartHeader handleClose={handleClose} numCartItems={numCartItems} />

        <CartList />

        {/* ---------- Animate CartFooter ---------- */}
        <motion.div
          variants={{
            animActive: { y: 0, x: 0, opacity: 1 },
            animReverse: { y: 300, x: 1000, opacity: 0 },
          }}
          initial={{ y: 300, x: 1000, opacity: 0 }}
          animate={cartAnimationClass}
          transition={{ type: 'tween', ease: 'easeOut', delay: 0.15 }}
        >
          <CartFooter numCartItems={numCartItems} goToCheckout={goToCheckout} />
        </motion.div>
        {/* ---------- End Animate CartFooter ----------* */}
      </SideCart>
    </>
  );
}

export default Cart;
