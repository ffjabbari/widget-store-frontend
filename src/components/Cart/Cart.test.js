/** Cart tests */

// local imports
import { renderWithStore } from '../../helpers/testHelpers';
import Cart from './Cart';

test('renders Cart', async () => {
  renderWithStore(<Cart />);
});

test('Cart snapshot', async () => {
  const { asFragment } = renderWithStore(<Cart />);
  expect(asFragment()).toMatchSnapshot();
});
