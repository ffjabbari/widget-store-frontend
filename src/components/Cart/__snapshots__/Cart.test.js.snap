// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-eCssSg fsMPpl ScreenBackground animActive"
  />
  <div
    class="sc-jSgupP edsasc SideCart animActive"
  >
    <div
      class="text-left px-4 pt-0 pt-lg-3"
    >
      <button
        class="btn-noStyle"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-window-close fa-w-16 fa-2x "
          data-icon="window-close"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-83.6 290.5c4.8 4.8 4.8 12.6 0 17.4l-40.5 40.5c-4.8 4.8-12.6 4.8-17.4 0L256 313.3l-66.5 67.1c-4.8 4.8-12.6 4.8-17.4 0l-40.5-40.5c-4.8-4.8-4.8-12.6 0-17.4l67.1-66.5-67.1-66.5c-4.8-4.8-4.8-12.6 0-17.4l40.5-40.5c4.8-4.8 12.6-4.8 17.4 0l66.5 67.1 66.5-67.1c4.8-4.8 12.6-4.8 17.4 0l40.5 40.5c4.8 4.8 4.8 12.6 0 17.4L313.3 256l67.1 66.5z"
            fill="currentColor"
          />
        </svg>
      </button>
      <span
        class="text-xxl pl-2 pl-sm-5"
      >
        Shopping Cart
      </span>
      <span
        class="CartIconBadge position-relative pl-4 pl-lg-5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-shopping-cart fa-w-18 fa-2x "
          data-icon="shopping-cart"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M528.12 301.319l47.273-208C578.806 78.301 567.391 64 551.99 64H159.208l-9.166-44.81C147.758 8.021 137.93 0 126.529 0H24C10.745 0 0 10.745 0 24v16c0 13.255 10.745 24 24 24h69.883l70.248 343.435C147.325 417.1 136 435.222 136 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-15.674-6.447-29.835-16.824-40h209.647C430.447 426.165 424 440.326 424 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-22.172-12.888-41.332-31.579-50.405l5.517-24.276c3.413-15.018-8.002-29.319-23.403-29.319H218.117l-6.545-32h293.145c11.206 0 20.92-7.754 23.403-18.681z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div>
      <div
        class="sc-bdfBwQ dTJGMw ItemsListDiv mx-auto text-dark"
      >
        <div
          class="sc-gsTCUz cBrQwO border-bottom py-1 bg-secondary text-light"
        >
          <div
            class="CartItem-name"
          >
            Item
          </div>
          <div
            class="CartItem-price"
          >
            Price
          </div>
          <div
            class="CartItem-quant-adj"
          >
            Quantity
          </div>
          <div
            class="CartItem-total"
          >
            Item Total
          </div>
        </div>
      </div>
      <div
        class="sc-dlfnbm kcZkaY"
      >
        <table
          class="sc-hKgILt iBFiPV TabulationTable ml-auto"
        >
          <tbody>
            <tr>
              <td
                class="text-left"
              >
                Subtotal
              </td>
              <td
                class="text-right"
              >
                <div>
                  $0
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      style="opacity: 0; transform: translateX(1000px) translateY(300px) translateZ(0);"
    >
      <div
        class="px-4 mt-lg-5"
      >
        <button
          class="w-100 mt-2 text-xl disabled btn btn-primary"
          type="button"
        >
          Proceed to Checkout
        </button>
      </div>
      <ul
        class="d-inline-block"
      >
        <li
          class="mt-3"
        >
          Tax and Shipping calculated at checkout.
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
